File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode a = head;
        while(n-->0){
            a = a.next;
        }
        if(a==null) return head.next;
        ListNode b = head;
        while(true){
            a=a.next;
            if(a==null) break;
            b=b.next;
        }
        if(b.next==null)return null;
        b.next = b.next.next;
        return head;
    }
}

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *front = head;
        while(n--){
            front = front->next;
        }
        ListNode *back = head;
        
        while(front && front->next){
            back = back->next;
            front = front->next;
        }
        if(front) back->next = back->next->next;
        else return head->next;
        return head;
    }
};